internal/docker/volumes.go

```diff
@@ -0,0 +1,166 @@
+package docker
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/volume"
+	client "github.com/docker/docker/client"
+)
+
+// A volume is a mounted directory on the homeserver container
+type Volume interface {
+	// Return the mount point
+	Mount() mount.Mount
+	// Prepare the mount point. `contextStr` is unique to this blueprint+homeserver for homeserver
+	// specific mounts e.g appservices.
+	Prepare(ctx context.Context, docker *client.Client, contextStr string) error
+}
+
+type VolumeCA struct {
+	source string
+	typ    mount.Type
+}
+
+// Prepare the Certificate Authority volume. This is independent of the homeserver calling Prepare
+// hence the contextual string is unused.
+func (v *VolumeCA) Prepare(ctx context.Context, docker *client.Client, x string) error {
+	// TODO: wrap in a lockfile
+	if os.Getenv("CI") == "true" {
+		// When in CI, Complement itself is a container with the CA volume mounted at /ca.
+		// We need to mount this volume to all homeserver containers to synchronize the CA cert.
+		// This is needed to establish trust among all containers.
+
+		containerID := getContainerID()
+		if containerID == "" {
+			return fmt.Errorf("failed to get container ID")
+		}
+		container, err := docker.ContainerInspect(ctx, containerID)
+		if err != nil {
+			return err
+		}
+		// Get the volume that matches the destination in our complement container
+		for i := range container.Mounts {
+			if container.Mounts[i].Destination == "/ca" {
+				v.source = container.Mounts[i].Name
+				v.typ = container.Mounts[i].Type
```

Comment:
It looks like we only set the `v.typ` when in CI which causes this issue, https://github.com/matrix-org/complement/pull/283/files#r783448724